c10/test/core/DispatchKeySet_test.cpp

```diff
@@ -3,25 +3,150 @@
 #include <unordered_set>
 
 #include <c10/core/DispatchKeySet.h>
+#include <c10/util/irange.h>
 
 using namespace c10;
 
+// This test exists not to be comprehensive, but to more clearly show
+// what the semantics of DispatchKeySet are.
+TEST(DispatchKeySet, ShowSemantics) {
```

Comment:
It's kind of confusing that we're calling the new thing a "DispatchKeySet" -- it is a set of DispatchKeys, but one can query weird things from them, like if "BackendBit::CPUBit" is included in the set ("BackendBit::CPUBit" is not a DispatchKey!), and certain sets are not actually representable (like {AutogradCUDA, CPU})

Reply:
Yeah, I think there's a good question here about the right way to model the whole thing. There are now 3 types of "keys/bits" that can validly go in a keyset:
- runtime keys (from the perspective of the dispatcher/DispatchKeySet these are all the same, although you could further divide these into "real runtime backends (CPU)", "fake backends" (FPGA), and "runtime functionalities" (Functionalize).
- "building block" functionality bits (Dense)
- "building block" backend bits (CPUBit).

And it would be pretty nice if we could model those 3 things with 3 separate enums. There are some problems with new enums / renaming things though:
- Because of BC, we're probably stuck with the existing names `DispatchKeySet` and DispatchKey`, and the existing set of dispatch keys.
- It's harder to add a separate enum for the "functionality building blocks", like Dense and Sparse. This is because there are existing priorities between "per-backend functionalities" and "normal functionalities" that we (probably?) need to maintain, like `Functionalize > Sparse > CustomRNGKeyId > Quantized > Meta`. We could probably add some side-car logic to know what the interleaved priorities are, but that seems pretty complicated. 

@bhosmer pointed out that modeling the backend bits with a second enum can help maintain some invariants at compile time (like, only backends can have a sparse/dense/autograd key... except that's a bad example because NestedTensor breaks that rule).

So that's why the current state of the PR includes a separate enum for backends, with everything else still in `DispatchKey`. I think there are probably two extremes here (only use the `DispatchKey` enum, and enforce all invariants about the different classes of keys at runtime, or use a separate enum for every type), and right now this PR tries to split them across different enums when possible.

That definitely has some cons though:
- it's maybe a little weird that you can ask a `DispatchKeySet` if it has either `DispatchKey`, or a `BackendBit`
- it's also a little weird that a `DispatchKey` may or may not map to an actual runtime key (this would be fixed if we decided to split off `Dense`/`Sparse`/`Quantized`/`AutogradFunctionality` into a separate enum).